File: backend/app/services/account_service.py
from datetime import datetime, timezone
from typing import List, Optional, Tuple
from sqlalchemy.orm import Session
from sqlalchemy import or_, desc, asc
from app.models.account import Account
from app.schemas.account import AccountCreate, AccountUpdate
from app.core.security import encrypt_credential, decrypt_credential, generate_nova_username

def get_next_nova_id(db: Session) -> str:
    """
    Generates non-random sequential ID: NOVA-0001, NOVA-0002, NOVA-0003...
    """
    last_account = db.query(Account).order_by(desc(Account.id)).first()
    if not last_account:
        next_num = 1
    else:
        next_num = last_account.id + 1
    return f"NOVA-{next_num:04d}"

def is_username_exists(db: Session, username: str, exclude_id: Optional[int] = None) -> bool:
    query = db.query(Account).filter(Account.username == username)
    if exclude_id:
        query = query.filter(Account.id != exclude_id)
    return query.first() is not None

def generate_unique_nova_username(db: Session) -> str:
    """Generates a Nova username and guarantees it is unique in database."""
    for _ in range(50):
        uname = generate_nova_username()
        if not is_username_exists(db, uname):
            return uname
    # Fallback with timestamp if extremely crowded
    import time
    return f"NovaSky{int(time.time()) % 10000}"

def create_account(db: Session, account_in: AccountCreate) -> Account:
    if is_username_exists(db, account_in.username):
        raise ValueError(f"Username '{account_in.username}' already exists.")

    nova_id = get_next_nova_id(db)
    encrypted_pwd = encrypt_credential(account_in.password)

    db_account = Account(
        nova_id=nova_id,
        username=account_in.username,
        encrypted_password=encrypted_pwd,
        status=account_in.status,
        notes=account_in.notes or "",
        created_at=datetime.now(timezone.utc),
        updated_at=datetime.now(timezone.utc)
    )
    db.add(db_account)
    db.commit()
    db.refresh(db_account)
    return db_account

def get_accounts(
    db: Session,
    query: Optional[str] = None,
    status: Optional[str] = None,
    sort_by: str = "newest"
) -> List[Account]:
    q = db.query(Account)

    if status and status.upper() != "ALL":
        q = q.filter(Account.status == status.upper())

    if query:
        search = f"%{query.strip()}%"
        q = q.filter(
            or_(
                Account.nova_id.ilike(search),
                Account.username.ilike(search),
                Account.notes.ilike(search)
            )
        )

    if sort_by == "oldest":
        q = q.order_by(asc(Account.id))
    elif sort_by == "username_asc":
        q = q.order_by(asc(Account.username))
    elif sort_by == "username_desc":
        q = q.order_by(desc(Account.username))
    else:  # newest
        q = q.order_by(desc(Account.id))

    return q.all()

def get_account_by_id(db: Session, account_id: int) -> Optional[Account]:
    return db.query(Account).filter(Account.id == account_id).first()

def get_account_credential(db: Session, account_id: int) -> Tuple[Account, str]:
    account = get_account_by_id(db, account_id)
    if not account:
        raise ValueError("Account not found.")
    
    # Touch last_used_at timestamp
    account.last_used_at = datetime.now(timezone.utc)
    db.commit()
    
    decrypted_password = decrypt_credential(account.encrypted_password)
    return account, decrypted_password

def update_account(db: Session, account_id: int, account_in: AccountUpdate) -> Account:
    account = get_account_by_id(db, account_id)
    if not account:
        raise ValueError("Account not found.")

    if account_in.username and account_in.username != account.username:
        if is_username_exists(db, account_in.username, exclude_id=account_id):
            raise ValueError(f"Username '{account_in.username}' already exists.")
        account.username = account_in.username

    if account_in.password:
        account.encrypted_password = encrypt_credential(account_in.password)

    if account_in.status:
        account.status = account_in.status

    if account_in.notes is not None:
        account.notes = account_in.notes

    account.updated_at = datetime.now(timezone.utc)
    db.commit()
    db.refresh(account)
    return account

def archive_account(db: Session, account_id: int) -> Account:
    """Soft delete by archiving the account."""
    account = get_account_by_id(db, account_id)
    if not account:
        raise ValueError("Account not found.")

    account.status = "ARCHIVED"
    account.updated_at = datetime.now(timezone.utc)
    db.commit()
    db.refresh(account)
    return account

def get_account_stats(db: Session) -> dict:
    total = db.query(Account).count()
    active = db.query(Account).filter(Account.status == "ACTIVE").count()
    paused = db.query(Account).filter(Account.status == "PAUSED").count()
    archived = db.query(Account).filter(Account.status == "ARCHIVED").count()
    return {"total": total, "active": active, "paused": paused, "archived": archived}

File: src/lib/account_service.ts
import { getAccountsData, getAccountStore, saveAccountsData, nowIso, type AccountRecord } from './db';
import { encryptCredential, decryptCredential } from './security';
import { generateNovaUsername } from './password';

export interface AccountRow extends AccountRecord {}

export interface AccountPublic {
  id: number;
  nova_id: string;
  username: string;
  status: string;
  tag: string;
  notes: string;
  created_at: string;
  updated_at: string;
  last_used_at?: string | null;
}

function rowToPublic(row: AccountRecord): AccountPublic {
  return {
    id: Number(row.id),
    nova_id: row.nova_id,
    username: row.username,
    status: row.status,
    tag: row.tag,
    notes: row.notes || '',
    created_at: row.created_at,
    updated_at: row.updated_at,
    last_used_at: row.last_used_at ?? null,
  };
}

function getAccountsFromStore(): AccountRecord[] {
  return getAccountsData();
}

export function getNextNovaId(): string {
  const accounts = getAccountsFromStore();
  const nextNum = accounts.reduce((max, account) => Math.max(max, Number(account.id) || 0), 0) + 1;
  return `NOVA-${String(nextNum).padStart(4, '0')}`;
}

export function isUsernameExists(username: string, excludeId?: number): boolean {
  return getAccountsFromStore().some((account) => account.username === username && account.id !== excludeId);
}

export function generateUniqueNovaUsername(): string {
  for (let i = 0; i < 50; i++) {
    const uname = generateNovaUsername();
    if (!isUsernameExists(uname)) return uname;
  }
  return `NovaSky${Date.now() % 10000}`;
}

export function createAccount(data: {
  username: string;
  password: string;
  status: string;
  notes?: string;
}): AccountPublic {
  if (isUsernameExists(data.username)) {
    throw new Error(`Username '${data.username}' already exists.`);
  }

  const accounts = getAccountsFromStore();
  const novaId = getNextNovaId();
  const enc = encryptCredential(data.password);
  const ts = nowIso();
  const newAccount: AccountRecord = {
    id: accounts.reduce((max, account) => Math.max(max, account.id), 0) + 1,
    nova_id: novaId,
    username: data.username,
    encrypted_password: enc,
    status: (data.status || 'ACTIVE').toUpperCase(),
    tag: 'GENERAL',
    notes: data.notes || '',
    created_at: ts,
    updated_at: ts,
    last_used_at: null,
  };

  accounts.push(newAccount);
  saveAccountsData(accounts);
  return getAccountById(newAccount.id)!;
}

export function getAccounts(opts: {
  query?: string | null;
  status?: string | null;
  sortBy?: string;
}): AccountPublic[] {
  const { query, status, sortBy = 'newest' } = opts;
  const accounts = getAccountsFromStore();

  const filtered = accounts.filter((account) => {
    if (status && status.toUpperCase() !== 'ALL' && account.status !== status.toUpperCase()) {
      return false;
    }
    if (query && query.trim()) {
      const like = query.trim().toLowerCase();
      if (!`${account.nova_id} ${account.username} ${account.notes}`.toLowerCase().includes(like)) {
        return false;
      }
    }
    return true;
  });

  const sorted = [...filtered].sort((a, b) => {
    if (sortBy === 'oldest') return a.id - b.id;
    if (sortBy === 'username_asc') return a.username.localeCompare(b.username);
    if (sortBy === 'username_desc') return b.username.localeCompare(a.username);
    return b.id - a.id;
  });

  return sorted.map(rowToPublic);
}

export function getAccountById(id: number): AccountPublic | null {
  const account = getAccountsFromStore().find((item) => item.id === id);
  return account ? rowToPublic(account) : null;
}

export function getAccountCredential(id: number): { account: AccountPublic; password: string } {
  const account = getAccountsFromStore().find((item) => item.id === id);
  if (!account) throw new Error('Account not found.');

  const updated = getAccountsFromStore().map((item) => (item.id === id ? { ...item, last_used_at: nowIso() } : item));
  saveAccountsData(updated);
  const password = decryptCredential(account.encrypted_password);
  return { account: rowToPublic(account), password };
}

export function updateAccount(
  id: number,
  data: { username?: string; password?: string; status?: string; notes?: string }
): AccountPublic {
  const existing = getAccountById(id);
  if (!existing) throw new Error('Account not found.');

  if (data.username && data.username !== existing.username && isUsernameExists(data.username, id)) {
    throw new Error(`Username '${data.username}' already exists.`);
  }

  const accounts = getAccountsFromStore();
  const updated = accounts.map((account) => {
    if (account.id !== id) return account;

    const newUsername = data.username ?? account.username;
    const newStatus = data.status ? data.status.toUpperCase() : account.status;
    const newNotes = data.notes !== undefined ? data.notes : account.notes;
    let enc = account.encrypted_password;

    if (data.password) {
      enc = encryptCredential(data.password);
    }

    return {
      ...account,
      username: newUsername,
      encrypted_password: enc,
      status: newStatus,
      notes: newNotes,
      updated_at: nowIso(),
    };
  });

  saveAccountsData(updated);
  return getAccountById(id)!;
}

export function archiveAccount(id: number): AccountPublic {
  const existing = getAccountById(id);
  if (!existing) throw new Error('Account not found.');

  const updated = getAccountsFromStore().map((account) =>
    account.id === id ? { ...account, status: 'ARCHIVED', updated_at: nowIso() } : account
  );
  saveAccountsData(updated);
  return getAccountById(id)!;
}

export function getAccountStats(): { total: number; active: number; paused: number; archived: number } {
  const accounts = getAccountsFromStore();
  return {
    total: accounts.length,
    active: accounts.filter((account) => account.status === 'ACTIVE').length,
    paused: accounts.filter((account) => account.status === 'PAUSED').length,
    archived: accounts.filter((account) => account.status === 'ARCHIVED').length,
  };
}

/** Export all accounts WITH decrypted passwords for backup. */
export function getAllAccountsWithPasswords(): { id: number; nova_id: string; username: string; password: string; status: string; notes: string; created_at: string }[] {
  return getAccountsFromStore().map((account) => ({
    id: Number(account.id),
    nova_id: account.nova_id,
    username: account.username,
    password: decryptCredential(account.encrypted_password),
    status: account.status,
    notes: account.notes || '',
    created_at: account.created_at,
  }));
}
